Various network stability fixes and reduce memory usage when saving settings

TD-er opened the pull request.

TD-er committed:
[stack usage] Allocate temp char array on heap for parsing arguments

These char arrays were rather large, so allocate on heap instead of char[240] on the stack.
Also re-order elements in not stored structs to reduce memory usage

TD-er committed:
Proper look for connection.

See discussion here: https://github.com/esp8266/Arduino/issues/5257

TD-er committed:
[Rules] Early exit on rules processing when rules not enabled

TD-er committed:
[#1910] Notification settings extra strings on stack

See #1910
Added a lot more of these changes and changed the input fields of the Notification Settings to be the same as all input fields in other places.

s0170071 commented on src/WebServer.ino:
```diff
@@ -1033,6 +1030,7 @@ void handle_config() {
   //String apkey = WebServer.arg(F("apkey"));
+  String ssid = WebServer.arg(F("ssid"));
 
 
```
Why not use the hasArg(F("ssid")) method ?

TD-er replied:
It is used later and the hasArg and arg method both do perform a search.
So it is a trade-off between memory and speed

clumsy-stefan commented:
tried this PR on my nodes (wemos&sonoff), they did to run fine until now, little more free memory but not more or less stable as before!

s0170071 commented:
@ clumsy-stefan What do you mean by no more stable? Did it crash?
On my end everything runs smoothly. No resets.

clumsy-stefan commented:
I do have random reboots on all nodes (before and after this patch) which are unrelated to this I assume as I use a recent esp core version from GIT and nt the standard build. so it could liekely be the reboots come from the core... that's why I said "Not more or less ;).... 

still trying to nail down the cause for the reboots I have, but as said as long as I use custom self built versions I won't blame the ESPEasy SW ;)

s0170071 commented:
I use 2.4.2 and do not have any issues. 
I use it for my sonoffs so there is no pwm nor any plugin involved.

s0170071 commented:
Are all your nodes on different power supplies?

TD-er commented:
Should we split the core 2.4.1 and 2.4.2 in separate builds for PWM and no time-critical PWM applications?
Or are there other issues with 2.4.2 which makes it better to stay at 2.4.1?
To me it looks like we may have tackled the stack issues now.

@s0170071 what LWIP version do you use?

clumsy-stefan commented:
yes, various differnt PS, but some of them with real big stable ones... and I already had uptimes of 3+ days with some versions...

btw: I use lwip2 low memory option latest version with features

btw2: in one of the latest commits from esp8266 core there is a change in how things are parsed (
https://github.com/esp8266/Arduino/pull/5252 and 
https://github.com/esp8266/Arduino/pull/5256 ) be aware that including these commits you won't be able to login anymore on the node, somehow the parsing of the password does not seem to work anymore.... did not yet find the time to analyze, I just reverted these two commits...

TD-er commented:
A few of my nodes are running core 2.4.2 and some have uptimes of around 7+ days.

And one outside running SDS011 sensor running build mega-20181001  
Uptime: 16 days 18 hours 31 minutes

s0170071 commented:
@TD-er  No. As I stated somewhere else already: lets look ahead. Use 2.4.2 as standard core. It has some issues fixed but unfortunately introduced another one. This bug is not ours and will ultimately sort itself out.  If someone has an issue with PWM, they can use an earlier version.

I use LWIPv2 "lower memory"

@clumsy-stefan 3 days is a start but not nearly enough. Do you self-compile with additional plugins, e.g. your own ? I found that adding plugins messes up the settings. You should factory reset then.

s0170071 commented:
@TD-er from my point of view things improved with

>  PR [#1906] Move Notification and Controller settings to heap

 . And I am pretty sure that this resolves some other issues too, especially those with a very long stack dump ;-)

clumsy-stefan commented:
@s0170071: yes self compiled with 2 or 3 additional plugins from the playground (eg. my own) and some which I disabled from the standard plugins... I use the same set for a long time and do factory resets from time to time... so I think that's not the problem... and as said, for sure I don't blame ESPEasy core currently, before pinning down where the issue lies...

PS: mostly it did not survive more than 3 days (about a month ago or so) as I always then integrated the newest commits from @TD-er & others (good work!!!) and therefore had to reboot the nodes ;)

s0170071 commented:
@clumsy-stefan please try this PR or PR1906. Before that the ESP was out of stack on various occasions. And stack issues show themselves as all sort of unpredictable behaviour. 
I found it good practice to use a fresh clean unit with no additional plugins as a reference. Set it aside and watch it. Maybe its your plugins ? No offence, but I experienced that myself. It crashed the ESP when there was a communications error - which only occurred once in a while.  

@TD-er did you introduce a delay somewhere ? My testing unit suddenly goes to modem sleep after all. If I let it alone, it consumes only 10mA (was 60mA two weeks ago)

clumsy-stefan commented:
@s0170071: thanks, both PR's are included already... and I agree, it could lead to issues.. that's why I^m trying to pin it down before complaining ;)

TD-er commented:
@s0170071 Yes, I added a few yields and maybe it is now also running more idle loops when rules are not activated.

s0170071 commented:
@TD-er test device appoaching 1 day continuous operation. No restarts.  Core 2.4.2 Lwip2 low

Free Mem: | 18960 (15872 - LoadNotificationSettings)
Free Stack: | 3520 (1600 - sendContentBlocking)

TD-er commented:
wow, 19k free. What plugins/controllers did you remove?
I have a test version running now, using a normal build + core 2.4.2/LWIP 1.4 and free mem is:

Free Mem: | 13632 (10376 - sendContentBlocking)
-- | --
Free Stack: | 3584 (1616 - sendHeaderBlocking)

s0170071 commented:
It's your pr, core 242 lwip2 and as far as I know no recently changed compiler flags. That's the extra ram that comes with cont stack moving to the sys area.
I have not removed plugins.

s0170071 commented:
1d 20h, still running, no memory leaks.

s0170071 commented:
Congrats @TD-er. I officially consider this  PR as the ultimate stability improvement.
 
Uptime: | 3 days 23 hours 15 minutes
Load: | 29.90% (LC=229)
Free Mem: | 18960 (15872 - LoadNotificationSettings)
Free Stack: | 3472 (1600 - sendContentBlocking)

I'll end this test now.

TD-er commented:
Great! :)
We're finally getting out of the twilight zone, where normal uptimes don't happen very often.

clumsy-stefan commented:
unfortunately with my self-compiled version my nodes currently do quite often reboot with software watchdog, hardware watchdog or exception. (every 20min, sometimes they just hang without even rebooting/reconnecting)..at some point in time I was also at 3+ days with my own version....

But I'm quite sure this is related to controller 009 I use, because when I disable the controller, I also have no reboots at all.... 
I think some combination of slow server, lots of node sending lots of data and filling the controller send queues still has an issue.

just as information, of course I have absolutely no expectation, for any support or similar (as self compiled etc.). however I try to pin it down, make it reproducable and will feedback if I have some specific issues/improvments...

that said, I also want to thank @TD-er (and all other) for the lot of work and time he invests in this project!!

TD-er commented:
@clumsy-stefan 
Maybe you can also test this PR: #1954

I replaced the calls to `yield()` with `delay(0)`.
My own test node, which reboots often with WDT reboots, is running a lot longer with this change.

N.B. I also added Sonoff POW R2 build to the PlatformIO in that PR.

clumsy-stefan commented:
yep, I saw that this afternoon and directly included and compiled it... the web interface feels a bit more snapy now... butunfortunately the nodes are not more stable...
only thing I'd appreciate is if you could just quickly review the current implementation of C009 (which are only a few lines) if you see any obvious issues (memory, something not initialized, memory not cleared, timeout not catched, etc)... but only if you have some spare minutes....

giig1967g commented:
@clumsy-stefan: i have reviewed the P009 plugin in #1942 .
There were some wrong variable type conversions.
Try #1942 and then tell me if it's better

TD-er commented:
@giig1967g 
@clumsy-stefan is referring to controller C009, not plugin P009.

giig1967g commented:
yes I noticed :(
I have deleted the post, but probably not fast enough...
Sorry

clumsy-stefan commented:
anyway, thanks for the help/efforts.... 👍  very kind!!!

one thing I subjectively feel is that enhancing the client timeout to 2000ms and reducing the min. sending interval to 10-20ms seem to improve stability/runtime.... but these are not yet facts, but probably can hint to something...

TD-er commented:
Reducing the minimum sending interval will for sure make the unit more responsive.
And it also may need less memory when sending a number of values in bursts.

TD-er commented:
I am looking now at C009

`element.valueNames[x] = ExtraTaskSettings.TaskDeviceValueNames[x];`

This can be taken out of the buffered elements I guess.
These are supposed to be the same always, so we can read them from the settings when actually sending the value.
Also, all the JSON objects are still present when it is converted to a string. I guess adding some scope may also reduce memory usage.
You maybe hit some kind of heap memory fragmentation?

clumsy-stefan commented:
first: thanks @TD-er for looking into it..

can I check heap fragmentation somehow?

Memory should not be a poblem, it's reportet every minute and stays stable at around 18k.. also Load looks ok.

PS: what I think is also strange is, that the reboot reasons vary (exception, SW-WD, HW-WD)

TD-er commented:
Exception is something interesting.
I will move the value names out of the queue elements.

TD-er commented:
@clumsy-stefan Maybe you can have a look at this PR:
https://github.com/letscontrolit/ESPEasy/pull/1957

Just changed it and the compiler is happy with it.
So I leave it to you to test.

clumsy-stefan commented:
thanks! I'm compiling as of now.... will report if it works...

clumsy-stefan commented:
no luck...
I think these changes make sense and should absolutely be included in the main branch!
however unfortunately they did not solve my problems...

I flashed 4 units (1 sonoff,, 2 wemos mini, 1 wemos pro), tow of them rebooted with a software watchdog, the other two with an exception. two of them stopped now responding (after they came up fine after the first reboot).

so I guess it's really underlying related to the esp core/IP stack... so I'll try to revert to some older version, unfortunately getting rid of some other improvements that have been made...

but again, thanks for your help, I think the changes should be included anyways as they run fine like this and use less memory...

clumsy-stefan commented:
I think I found the problem with my nodes!!!

It's FULL MOON !! too much radiation!!

I'm sure from sunday on they will work again!!

🤣 🤣 

(sorry for spamming, but I needed to get rid of that, I'm getting a bit desperate currently)..

clumsy-stefan commented:
@TD-er: I think I pinned down the network issues. As I'm using esp core, they updated about 3 weeks ago to the latest Nonos SDK V3.0.0 this seem to cause the various freezes where the nodes not even reconnect to the wifi and just freeze.
returning to the 2.2.1 SDK and lwip 2.0.1 improved at least that the nodes "just" reboot with a software watchdog and reconenct to the wlan..

at least I will then be able to help when you decide to move to the new SDK/LWIP/Nonos, as I have quite some experience now ;)

again thanks for your help!!
